dockstore-webservice/src/main/java/io/openapi/api/impl/ToolsApiServiceImpl.java

```diff
@@ -360,45 +359,62 @@ public Response toolsGet(String id, String alias, String toolClass, String descr
-                    Joiner.on('&').join(filters) + "&offset=" + (offsetInteger + 1), null).normalize();
-                responseBuilder.header("next_page", nextPageURI.toURL().toString());
-            }
-            URI lastPageURI = new URI(config.getExternalConfig().getScheme(), null, config.getExternalConfig().getHostname(), port,
-                ObjectUtils.firstNonNull(config.getExternalConfig().getBasePath(), "") + relativePath, Joiner.on('&').join(filters) + "&offset=" + numPages, null)
-                .normalize();
-            responseBuilder.header("last_page", lastPageURI.toURL().toString());
+        responseBuilder.header("self_link", createUrlString(scheme, hostname, port, path, encodedQuery));
+        if (startIndex + actualLimit < numEntries.sum()) {
+            responseBuilder.header("next_page", createUrlString(scheme, hostname, port, path, positionQuery(encodedQuery, actualLimit, offsetInteger + 1L)));
+        }
+        final long numPages = numEntries.sum() / actualLimit;
+        responseBuilder.header("last_page", createUrlString(scheme, hostname, port, path, positionQuery(encodedQuery, actualLimit, numPages)));
 
+        return responseBuilder.build();
+    }
+
+    private String createUrlString(String scheme, String hostname, int port, String path, String encodedQuery) {
+        String url;
+        try {
+            url = new URI(scheme, null, hostname, port, path, null, null).normalize().toURL().toString();
         } catch (URISyntaxException | MalformedURLException e) {
-            throw new CustomWebApplicationException("Could not construct page links", HttpStatus.SC_BAD_REQUEST);
+            throw new CustomWebApplicationException("Could not create url string", HttpStatus.SC_BAD_REQUEST);
+        }
+        // The URI constructor tries to encode the query string that it is
+        // passed, and there's no way to stop it, so we add it here instead.
+        if (encodedQuery != null) {
+            url += '?';
+            url += encodedQuery;
+        }
+        return url;
+    }
+
+    private String positionQuery(String encodedQuery, long limit, long offset) {
+        // For more sophisticated query string processing, the
+        // https://hc.apache.org/httpcomponents-client-5.1.x/
+        // library may be of use.
+        List<String> resultNameValues = new ArrayList<>();
+        // Split the query string into name=value pairs at the
+        // ampersands, then copy each name=value pair unchanged, except
+        // for the "limit" and "offset" pairs, which we omit. We will
+        // add them back at the end.
+        if (encodedQuery != null) {
+            for (String nameValue: encodedQuery.split("&")) {
+                if (!nameValue.startsWith("limit=") && !nameValue.startsWith("offset=")) {
+                    resultNameValues.add(nameValue);
+                }
+            }
         }
-        trsListener.loadTRSResponse(hashcode, responseBuilder);
```

Comment:
Hmmm, should have noticed this earlier. 
I only see this on the left side, does this break caching? 
(I know we're removing it in develop ... )

Reply:
Oh shoot, I had originally targeted this PR to develop, it broke when I retargeted to the hotfix. :(

Reply:
Fixed.